crates/holochain_cascade/src/authority/get_agent_activity_query/must_get_agent_activity.rs

```diff
@@ -1,157 +1,223 @@
+        .collect();
 
-    // For all the hashes in the filter, get their sequences.
-    Sequences::find_sequences(filter, get_seq_from_hash, get_seq_from_ts)
+    Ok(warrants)
 }
 
-/// Get the agent activity for a given range of actions
-/// from the database.
-fn get_activity(
+/// Get the agent activity for a given range of actions from the database.
+pub(crate) fn get_filtered_agent_activity(
     txn: &Transaction,
-    scratch: Option<&Scratch>,
     author: &AgentPubKey,
-    range: &RangeInclusive<u32>,
+    filter: ChainFilter,
+    filter_chain_top_action_seq: u32,
 ) -> StateQueryResult<Vec<RegisterAgentActivity>> {
-    let mut out = txn
+    // Get the max action seq of all Actions in the set of until hashes.
+    let chain_filter_limit_conditions_until_hashes_max_seq = if let Some(filter_hashes) =
+        filter.get_until_hash()
+    {
+        // Construct sql query with placeholders for list elements.
+        //
+        // We cannot keep this sql query in a standalone file,
+        // because at compile-time we don't know how many '?' placeholders to include.
+        let filter_hashes_placeholder = filter_hashes
+            .iter()
+            .map(|_| "?")
+            .collect::<Vec<&str>>()
+            .join(", ");
+        let sql_query_seq_hash_in_set = format!(
+            "
+            SELECT
+                MAX(Action.seq) as max_action_seq
+            FROM
+                Action
+                JOIN DhtOp ON DhtOp.action_hash = Action.hash
+            WHERE
+                Action.hash IN ({filter_hashes_placeholder})
+                AND Action.author = ?
+                AND DhtOp.type = ?
+                AND DhtOp.when_integrated IS NOT NULL
+        "
+        );
+
+        // Prepare query parameters
+        let mut query_params: Vec<Box<dyn ToSql>> = filter_hashes
+            .iter()
+            .map(|h| -> Box<dyn ToSql> { Box::new(h.clone()) })
+            .collect();
+        query_params.push(Box::new(author));
+        query_params.push(Box::new(ChainOpType::RegisterAgentActivity));
+
+        let query_params_refs: Vec<&dyn ToSql> = query_params.iter().map(|v| v.as_ref()).collect();
+        let query_params_refs_slice: &[&dyn ToSql] = query_params_refs.as_slice();
+
+        // Execute query
+        let max_action_seq: Option<u32> = txn
+            .prepare(&sql_query_seq_hash_in_set)?
+            .query_row(query_params_refs_slice, |row| row.get("max_action_seq"))?;
+
+        max_action_seq
+    } else {
+        None
+    };
+
+    // Until hash Action seq must be less than or equal to ChainFilter chain_top Action seq
+    if let Some(until_action_seq) = chain_filter_limit_conditions_until_hashes_max_seq {
+        if until_action_seq > filter_chain_top_action_seq {
+            return Err(StateQueryError::InvalidInput("The largest ChainFilter until hash Action seq must be less than or equal to the ChainFilter chain_top action seq.".to_string()));
```

Comment:
Casing even more critical in error messaging. It's `LimitConditions::UntilHash` that needs to be referenced here I think?

Both because that's easier for a user to investigate but also because tooling is more likely to notice a rename later if the referencing is crystal clear.